### PR TITLE
Disallow internal function types as parameters for public/external library function

### LINKED ISSUE #6151: Internal functions can be passed in library calls
A struct containing an internal function can be passed across library calls. This is non-sensical, since the library could then call the internal function whose code is not even available.

Example code:

```
library L {
    struct S {
        function() internal x;
    }

    function f(S storage s) public {
        s.x();
    }
}

contract c{
    L.S s;
    function f() public {
        s.x = f;
        L.f(s);
    }
}
```

### LINKED ISSUE #6215: Internal Error when passing array of internal-function-refs to library function
```
library L {
	function a(function() internal[] storage ff) { }
}
```
```
Internal compiler error during compilation:
/home/marenz/ethereum/solidity/libsolidity/ast/Types.cpp(2765): Throw in function virtual std::__cxx11::string dev::solidity::FunctionType::canonicalName() const
Dynamic exception type: boost::wrapexcept<langutil::InternalCompilerError>
std::exception::what:
[dev::tag_comment*] =
```

### PR DESCRIPTION
**Pull request by Marenz**
I am not very sure about my changes here, but the tests all work, so.. yeah.

Fixes #6215 and fixes #6151

### CODE REVIEW

File: test/libsolidity/syntaxTests/functionTypes/internal_function_array_as_external_parameter_in_library_external.sol
Comment (chriseth) [inline test/libsolidity/syntaxTests/functionTypes/internal_function_array_as_external_parameter_in_library_external.sol]: Do we have a test that uses `storage` here?

**Comment by chriseth** [comment]
Looks good! I think MappingType also needs a similar change.

File: test/libsolidity/syntaxTests/functionTypes/internal_function_array_as_external_parameter_in_library_external.sol
Comment (Marenz) [reply test/libsolidity/syntaxTests/functionTypes/internal_function_array_as_external_parameter_in_library_external.sol]: I'll add one

**Comment by Marenz** [comment]
Hmm are you sure? This looks okay to me

```
	TypePointer interfaceType(bool _inLibrary) const override
	{
		return _inLibrary ? shared_from_this() : TypePointer();
	}
```

**Comment by chriseth** [comment]
The check for mappings does not descend into the value type, it just checks for storage or non-storage.

**Comment by Marenz** [comment]
Ah, right. Of course.

**Comment by Marenz** [comment]
But mappings can't be passed to external functions anyway?

File: libsolidity/ast/Types.cpp
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: I just created #6223, because this is in fact horrible IMO...

Comment (Marenz) [reply libsolidity/ast/Types.cpp]: I agree. I was just too much focused on "How the hell is this supposed to work" to have any capacity left for "How could this work instead" ;)

Comment (ekpyron) [reply libsolidity/ast/Types.cpp]: Yeah - fixing this is a job for a separate PR anyways :-).

File: libsolidity/ast/Types.cpp
Comment (ekpyron) [inline libsolidity/ast/Types.cpp]: This is only half of what's happening - I think you should keep the first line of the original comment.

**Comment by ekpyron** [comment]
@Marenz 
```
library L {
    function f(mapping(uint => function(uint) internal returns (uint)) storage m) public {
    }
}
```
-> InternalCompilerError

Mappings *can* be passed to public library functions - such a public library function is called with "delegatecall", when called externally and such a call keeps the caller's storage - that's why they can have mappings as arguments, since those live in storage and for delegatecall's we still have the right storage on the callee side.

**Comment by Marenz** [comment]
I pushed an update so now the mapping type also travels into the types

**Comment by Marenz** [comment]
rebased on latest

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: I think recursive storage structs should still be allowed inside libraries.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: could call `canBeUsedExternally` - but I guess we will soon get rid of that anyway...

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: I'm actually surprised that we do not seem to have a test for this...

**Comment by Marenz** [comment]
Added tests for recursion and adjusted code to allow recursive types in public library functions again

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: I think this can be an assertion - keyType has to be an elementary type.

**Comment by Marenz** [comment]
Updated:
* changed if to assert
* removed "recursive()"
* caching result of call in StructType

File: Changelog.md
Comment (chriseth) [inline Changelog.md]: This looks like a rebase error.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: This line is reached for non-recursive structs - is that the intention?

**Comment by chriseth** [comment]
Does this also fix https://github.com/ethereum/solidity/issues/4260 ?

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: Yes, for "normal" (non-recursive) structs we still have to check each member whether it can be used externally

**Comment by Marenz** [comment]
> Does this also fix #4260 ?

I just tested the provided code and it still gives that error message, so no, it doesn't

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: But this will again run the cycle detector (a different instance, though) that we just ran on that struct, right? Even worse: What happens if `_inLibrary` is true - won't this lead to an infinite loop?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Can we determine the value for both at the same time? Isn't it basically the same work?

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: This line is only reached if it's not a struct (or array of structs) that somewhere contains the current struct due to the check in line 2184. If the `if` gets a result it will either skip the member and look at the next (`_inLibrary => continue`) or stop the loop altogether (`return;`)

Comment (Marenz) [reply libsolidity/ast/Types.cpp]: So, it might run a cycle detector again if this is a struct, but at that point it can not be a struct that contains this struct.

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: Well, not exactly. If it's `!_inLibrary` it stops when it detects a recursion thus the check for internal function types is not run on the whole tree.

Comment (Marenz) [reply libsolidity/ast/Types.cpp]: I suppose in the case of `inLibrary == true` we can set both

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Ah sorry, I somehow did not recognize that the `continue` reaches that far outside :)

**Comment by Marenz** [comment]
Updated:
* now also caching the result for the `!_inLibrary` case if we detect that it's recursive

File: libsolidity/analysis/TypeChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: If I want to pass a recursive struct in memory to a library, it should not be recursive either - isn't it?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp:2148]: Yeah, I think whether or not the location is storage is not considered anymore.

Comment (Marenz) [reply libsolidity/ast/Types.cpp:2148]: It wasn't always considered before either though, e.g. if the location is memory, _inlibrary is true but all the types can be used externally in the struct. How should the new behavior be?

Comment (chriseth) [reply libsolidity/ast/Types.cpp:2148]: The old behaviour was: If the location is not storage, `_inLibrary` is ignored.
The new behaviour:
 - disallow internal function pointers regardless of storage location or library-flag
 - otherwise identical with old behaviour, i.e.:

- allow recursive structs if location is storage and library-flag is set.
- allow mapping members if location is storage and library-flag is set

**Comment by Marenz** [comment]
This PR is on hold until #6223 is implemented because printing the correct error message for the case
```
    struct MyStructName {
        address addr;
        MyStructName[] x;
    }

    function f(MyStructName memory _x) public {
	}
}
```

is currently very complicated as the error message right now only depends on whether `interfaceType()` returns true/false and whether`isLibrary` is set. 

The above should fail because the data location is not `storage` which is required for recursive structs.

**Comment by stackenbotten** [comment]
There was an error when running `test_check_style` for commit `6a5ec7460c2d4db6220ee66640a5bb04ae01b4f8`:
```
Coding style error:
 libsolidity/ast/Types.cpp:3273: return TypeResult::err("Only libraries are allowed to have the mapping type in public or external functions.");

```
Please check that your changes are working as intended.

**Comment by Marenz** [comment]
uuupdated:
* interfaceType now uses `TypeResult` return type so it can return errors
* the various error situations now all have dedicated error messages

File: Changelog.md
Comment (erak) [inline Changelog.md]:
```suggestion
 * Type System: ....
```

File: libsolidity/ast/Types.cpp
Comment (erak) [inline libsolidity/ast/Types.cpp]: Additional `;`

File: libsolidity/ast/Types.cpp
Comment (erak) [inline libsolidity/ast/Types.cpp]: Mmh, this looks a bit odd. Maybe `TypeResult{TypePointer{}}`?

File: libsolidity/ast/Types.cpp
Comment (erak) [inline libsolidity/ast/Types.cpp:2276]: Ah, that's unfortunate. I'd probably need this function for https://github.com/ethereum/solidity/pull/6282. Is it possible to split it out again?

Comment (Marenz) [reply libsolidity/ast/Types.cpp:2276]: Can you just use `interfaceType()` directly? It kinda does that check implicitly

Comment (Marenz) [reply libsolidity/ast/Types.cpp:2276]: I could offer a dedicated function that uses `interfaceType()` to gather that information I suppose

Comment (Marenz) [reply libsolidity/ast/Types.cpp:2276]: like, `interfaceType()` would set an `boost::optional<bool> m_recursive;`

Comment (erak) [reply libsolidity/ast/Types.cpp:2276]: Having a flag that signals if this struct declares a member of the same type e.g. `boost::optional<bool> m_recursive` would be enough I guess.

**Comment by Marenz** [comment]
Addressed eriks comments.

Re-added recursive() which internally just uses interfaceType()

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp:2247]: This `.get()` should not be necessary, should it?

File: libsolidity/analysis/TypeChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Where does `"Invalid type! ()"` appear? Shouldn't it be an assertion? If not, what is the idea behind the `()`?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: What does `Ext` refer to?

File: libsolidity/analysis/TypeChecker.cpp
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: I forgot to take that out, it was merely to differentiate this from another (equally phrased) error. We can probably turn it into an assert now, yes.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Isn't `recursion` always false prior to this line? It could be declared at this line, actually.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: `Return false` is not really accurate anymore.

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: I just re-orded & adjusted the existing code which already had & used `baseExt`, so.. I actually have no idea. Something related to `External` maybe? because.. external type?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: "Recursive structs can only be passed as storage pointers to libraries, not as memory objects to contract functions.'

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Ah, it cannot be declared here, but I think the `||` is not needed.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp:2189]: This will run the whole code above another time - have you tried eliminating this somehow?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Please swap the two cases to eliminate the negation.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: No braces needed.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: "have the mapping type"?

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp:2247]: It is indeed required:
```
Types.cpp:2245:12: Fehler: Basisoperand von »->« hat Nicht-Zeiger-Typ »dev::Result<std::shared_ptr<const dev::solidity::Type> >«
    return t->signatureInExternalFunction(_structsByName);
            ^~
```

File: test/libsolidity/syntaxTests/structs/recursion/recursive_struct_as_memory_library_function_parameter.sol
Comment (chriseth) [inline test/libsolidity/syntaxTests/structs/recursion/recursive_struct_as_memory_library_function_parameter.sol]: Indentation

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Ah sorry, I didn't see that. Perhaps `baseInterfaceType` would be a better name?

Comment (chriseth) [reply libsolidity/ast/Types.cpp]: I think the function was called `externalType` before.

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: it's also assigned to in line 2176

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp:2189]: But it also needs to run the code above? I mean, we could have two different structs both recursive for each other and also embedded into each other? e.g.

```
struct A 
{
    A[] array;
    uint val;
};

struct B
{
    A[] a_array;
    uint x;
    B[] b_array;
}
```

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: "use" instead?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp:2189]: Right, the recursion detection has to break at a different point.

Comment (chriseth) [reply libsolidity/ast/Types.cpp:2189]: Since `recursive` is now a member, perhaps we could break the cycle detector early if one of the members is already known to be recursive?

File: libsolidity/ast/Types.cpp
Comment (chriseth) [inline libsolidity/ast/Types.cpp]:
```suggestion
				if (innerStruct->m_recursive == true || _cycleDetector.run(innerStruct->structDefinition()))
```

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp:2189]: But we still need to check all the other members if recursion is allowed in the library case..
In the other case we already break the cycle as soon as we discover one.

File: libsolidity/ast/Types.cpp
Comment (Marenz) [reply libsolidity/ast/Types.cpp]: ah ,that's what you meant. Got it. :+1:

**Comment by Marenz** [comment]
Addressed the comments.
Will squash once approved.

File: libsolidity/ast/Types.cpp
Comment (chriseth) [reply libsolidity/ast/Types.cpp]: Can you add this? Otherwise the PR is fine.

Comment (Marenz) [reply libsolidity/ast/Types.cpp]: I thought I did but it turned out I didn't save the file :1st_place_medal:

**Comment by Marenz** [comment]
should be ready now :>

**Review by chriseth — APPROVED** [review]

**Pull request merged**
